### PR TITLE
[API]Product image handling

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         | master 
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

**Review by GSadee — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -33,6 +33,11 @@ public function getConfigTreeBuilder(): TreeBuilder
                     ->defaultFalse()
                 ->end()
             ->end()
+            ->children()
+                ->variableNode('product_image_prefix')
+                    ->defaultValue('media/image')
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/DependencyInjection/Configuration.php:38]: I'm fine with defining this value as a parameter, however we should at least use another parameter, that will be used here as well: https://github.com/Sylius/Sylius/blob/5dd517bb2198409d3f5723dfef1c290ec3f991e5/src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml#L60

But once we will introduce the parameter there, I'm not sure we need it. Up to you.

File: src/Sylius/Bundle/ApiBundle/Serializer/ProductImageNormalizer.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        if (isset($context[self::ALREADY_CALLED])) {
+            return false;
+        }
+
+        return $data instanceof ProductImageInterface;
+    }
+
+    private function validatePrefix(string $prefix): string
+    {
+        if ('/' !== substr($prefix, 0, 1)) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Serializer/ProductImageNormalizer.php:62]: lets use directory separator constant 
```suggestion
        if (\DIRECTORY_SEPARATOR !== substr($prefix, 0, 1)) {
```

File: src/Sylius/Bundle/ApiBundle/Serializer/ProductImageNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    public function normalize($object, $format = null, array $context = [])
+    {
+        Assert::isInstanceOf($object, ProductImageInterface::class);
+        Assert::keyNotExists($context, self::ALREADY_CALLED);
+
+        $context[self::ALREADY_CALLED] = true;
+
+        $data = $this->normalizer->normalize($object, $format, $context);
+
+        $data['media_path'] = $this->prefix . $data['path'];
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Serializer/ProductImageNormalizer.php:46]: Is `media_path` added to the endpoint documentation? Nonetheless, I would remove this additional field and just replace the value of `path`. Without prefix it is useless.
```suggestion
        $data['path'] = $this->prefix . $data['path'];
```

File: src/Sylius/Bundle/ApiBundle/spec/Serializer/ProductImageNormalizerSpec.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+use Sylius\Component\Core\Model\ProductImageInterface;
+use Symfony\Component\Serializer\Normalizer\ContextAwareNormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+final class ProductImageNormalizerSpec extends ObjectBehavior
+{
+    function let(): void
+    {
+        $this->beConstructedWith('prefix');
+    }
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/Serializer/ProductImageNormalizerSpec.php:27]: Constructor has 4 different paths that can be executed, only the first if is tested. We should cover 3 additional cases:
* `/prefix`
* `/prefix/`
* `prefix/`

File: UPGRADE-API-1.11.md [hunk 5]
```diff
@@ -1,5 +1,27 @@
...
+   
+    ```diff
+        {
+            "@context": "/api/v2/contexts/ProductImage",
+            "@id": "/api/v2/shop/product-images/123",
+            "@type": "ProductImage",
+            "id": "123",
+            "type": "thumbnail",
+            "path": "uo/product.jpg",
+    +       "media_path": "/media/image/uo/product.jpg"
```
Comment (GSadee) [inline UPGRADE-API-1.11.md:14]: If I see correctly, there is no more `media_path`, you are updating `path` field

File: src/Sylius/Bundle/ApiBundle/Serializer/ProductImageNormalizer.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        if (\DIRECTORY_SEPARATOR !== substr($prefix, 0, 1)) {
+            $prefix = '/'.$prefix;
+        }
+
+        if (\DIRECTORY_SEPARATOR === substr($prefix, -1)) {
+            return $prefix;
+        }
+
+        return $prefix . '/';
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Serializer/ProductImageNormalizer.php:70]:
```suggestion
        if (\DIRECTORY_SEPARATOR !== substr($prefix, 0, 1)) {
            $prefix = \DIRECTORY_SEPARATOR . $prefix;
        }

        if (\DIRECTORY_SEPARATOR === substr($prefix, -1)) {
            return $prefix;
        }

        return $prefix . \DIRECTORY_SEPARATOR;
```
?

File: UPGRADE-API-1.11.md (hunk 5 above)
Comment (arti0090) [reply UPGRADE-API-1.11.md]: I have changed it a little now :)

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, @arti0090! :tada:

**Pull request merged**